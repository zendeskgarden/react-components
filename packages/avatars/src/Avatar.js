/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import PropTypes from 'prop-types';
import styled from 'styled-components';
import classNames from 'classnames';
import { retrieveTheme } from '@zendeskgarden/react-theming';
import AvatarStyles from '@zendeskgarden/css-avatars';

const COMPONENT_ID = 'avatars.avatar';
const SIZE = {
  SMALL: 'small',
  LARGE: 'large'
};

/**
 * Accepts all `<figure>` props
 */
const Avatar = styled.figure.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION,
  className: props =>
    classNames(AvatarStyles['c-avatar'], {
      // Types
      [AvatarStyles['c-avatar--system']]: props.system,

      // Sizes
      [AvatarStyles['c-avatar--small']]: props.size === SIZE.SMALL,
      [AvatarStyles['c-avatar--large']]: props.size === SIZE.LARGE,

      // State
      [AvatarStyles['is-disabled']]: props.disabled,
      [AvatarStyles['c-avatar--borderless']]: props.isBorderless
    })
})`
  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

Avatar.propTypes = {
  /** Applies system styling */
  system: PropTypes.bool,
  size: PropTypes.oneOf([SIZE.SMALL, SIZE.LARGE]),
  disabled: PropTypes.bool,
  /** Removes border for improved display on a dark background */
  isBorderless: PropTypes.bool
};

/** @component */
export default Avatar;
