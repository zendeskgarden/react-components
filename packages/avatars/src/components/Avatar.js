/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { Children } from 'react';
import PropTypes from 'prop-types';
import { StyledAvatar, StyledText } from '../styled';

const SIZE = {
  EXTRASMALL: 'extrasmall',
  SMALL: 'small',
  MEDIUM: 'medium',
  LARGE: 'large'
};

const STATUS = {
  AVAILABLE: 'available',
  ACTIVE: 'active',
  AWAY: 'away'
};

/**
 * Accepts all `<figure>` attributes and events
 */
const Avatar = React.forwardRef(
  (
    {
      isSystem,
      size,
      status,
      children,
      badge,
      surfaceColor,
      backgroundColor,
      foregroundColor,
      ...other
    },
    ref
  ) => {
    const computedStatus = badge === undefined ? status : STATUS.ACTIVE;

    return (
      <StyledAvatar
        ref={ref}
        isSystem={isSystem}
        size={size}
        status={computedStatus}
        data-badge={badge}
        surfaceColor={surfaceColor}
        backgroundColor={backgroundColor}
        foregroundColor={foregroundColor}
        aria-atomic="true"
        aria-live="polite"
        {...other}
      >
        {Children.only(children)}
      </StyledAvatar>
    );
  }
);

Avatar.propTypes = {
  /** Set the avatar background color */
  backgroundColor: PropTypes.string,
  /** Set the color for child SVG or `Avatar.Text` components */
  foregroundColor: PropTypes.string,
  /** Set the color of the surface behind the avatar â€“ used to manipulate the inner status rings */
  surfaceColor: PropTypes.string,
  /** Applies system styling */
  isSystem: PropTypes.bool,
  badge: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  size: PropTypes.oneOf([SIZE.EXTRASMALL, SIZE.SMALL, SIZE.MEDIUM, SIZE.LARGE]),
  status: PropTypes.oneOf([STATUS.AVAILABLE, STATUS.AWAY]),
  children: PropTypes.node
};

Avatar.defaultProps = {
  size: SIZE.MEDIUM
};

Avatar.Text = StyledText;

/** @component */
export default Avatar;
