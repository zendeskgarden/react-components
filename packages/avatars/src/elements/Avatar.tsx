/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { Children, HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledAvatar, StyledText } from '../styled';

interface IAvatarProps extends HTMLAttributes<HTMLElement> {
  /** Sets the avatar background color */
  backgroundColor?: string;
  /** Sets the color for child SVG or `Avatar.Text` components */
  foregroundColor?: string;
  /** Provides surface color for an avatar placed on a non-white background */
  surfaceColor?: string;
  /** Applies system styling for representing objects, brands, or products */
  isSystem?: boolean;
  /** Specifies the avatar size */
  size?: 'extraextrasmall' | 'extrasmall' | 'small' | 'medium' | 'large';
  /** Applies status styling */
  status?: 'available' | 'away';
  /** Sets the badge text and applies active styling */
  badge?: string | number;
}

/**
 * @extends HTMLAttributes<HTMLElement>
 */
const Avatar: React.FunctionComponent<IAvatarProps> = React.forwardRef<HTMLElement, IAvatarProps>(
  (
    {
      isSystem,
      size,
      status,
      children,
      badge,
      surfaceColor,
      backgroundColor,
      foregroundColor,
      ...other
    },
    ref
  ) => {
    const computedStatus = badge === undefined ? status : 'active';

    return (
      <StyledAvatar
        ref={ref}
        isSystem={isSystem}
        size={size}
        status={computedStatus}
        data-badge={badge}
        surfaceColor={surfaceColor}
        backgroundColor={backgroundColor}
        foregroundColor={foregroundColor}
        aria-atomic="true"
        aria-live="polite"
        {...other}
      >
        {Children.only(children)}
      </StyledAvatar>
    );
  }
);

Avatar.displayName = 'Avatar';

Avatar.propTypes = {
  backgroundColor: PropTypes.string,
  foregroundColor: PropTypes.string,
  surfaceColor: PropTypes.string,
  isSystem: PropTypes.bool,
  badge: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  size: PropTypes.oneOf(['extraextrasmall', 'extrasmall', 'small', 'medium', 'large']),
  status: PropTypes.oneOf(['available', 'away'])
};

Avatar.defaultProps = {
  size: 'medium'
};

(Avatar as any).Text = StyledText;

export default Avatar as React.FunctionComponent<
  IAvatarProps & React.RefAttributes<HTMLElement>
> & {
  Text: typeof StyledText;
};
