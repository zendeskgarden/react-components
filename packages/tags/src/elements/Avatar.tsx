/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { SVGAttributes } from 'react';
import { StyledAvatar } from '../styled';

const AvatarComponent = (props: SVGAttributes<SVGElement>) => <StyledAvatar {...props} />;

AvatarComponent.displayName = 'Tag.Avatar';

/**
 * @extends SVGAttributes<SVGElement>
 */
export const Avatar = AvatarComponent;
