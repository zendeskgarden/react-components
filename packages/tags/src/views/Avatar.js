/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { Children } from 'react';
import classNames from 'classnames';
import TagStyles from '@zendeskgarden/css-tags';

const COMPONENT_ID = 'tags.avatar';

const Avatar = ({ children, className }) => {
  return React.cloneElement(Children.only(children), {
    'data-garden-id': COMPONENT_ID,
    'data-garden-version': PACKAGE_VERSION,
    className: classNames(className, TagStyles['c-tag__avatar'])
  });
};

/** @component */
export default Avatar;
