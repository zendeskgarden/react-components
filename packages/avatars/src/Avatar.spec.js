/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { shallow } from 'enzyme';
import Avatar from './Avatar';

describe('Avatar', () => {
  const defaultImage = <img src="test" alt="test" />;

  it('applies default styling correctly', () => {
    const wrapper = shallow(<Avatar>{defaultImage}</Avatar>);

    expect(wrapper).toMatchSnapshot();
  });

  it('applies system styling correclty if provided', () => {
    const wrapper = shallow(<Avatar system>{defaultImage}</Avatar>);

    expect(wrapper).toMatchSnapshot();
  });

  describe('Sizes', () => {
    ['small', 'large'].forEach(size => {
      it(`applies ${size} correctly if provided`, () => {
        const wrapper = shallow(<Avatar size={size}>{defaultImage}</Avatar>);

        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('States', () => {
    it('applies disabled styling if provided', () => {
      const wrapper = shallow(<Avatar disabled>{defaultImage}</Avatar>);

      expect(wrapper).toMatchSnapshot();
    });

    it('applies borderless styling if provided', () => {
      const wrapper = shallow(<Avatar isBorderless>{defaultImage}</Avatar>);

      expect(wrapper).toMatchSnapshot();
    });
  });
});
