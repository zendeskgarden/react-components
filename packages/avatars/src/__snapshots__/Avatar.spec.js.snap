// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar Sizes applies large correctly if provided 1`] = `
<figure
  className="c-avatar c-avatar--large "
  data-garden-id="avatars.avatar"
  data-garden-version="version"
  size="large"
>
  <img
    src="test"
  />
</figure>
`;

exports[`Avatar Sizes applies small correctly if provided 1`] = `
<figure
  className="c-avatar c-avatar--small "
  data-garden-id="avatars.avatar"
  data-garden-version="version"
  size="small"
>
  <img
    src="test"
  />
</figure>
`;

exports[`Avatar States applies borderless styling if provided 1`] = `
<figure
  className="c-avatar c-avatar--borderless "
  data-garden-id="avatars.avatar"
  data-garden-version="version"
>
  <img
    src="test"
  />
</figure>
`;

exports[`Avatar States applies disabled styling if provided 1`] = `
<figure
  className="c-avatar is-disabled "
  data-garden-id="avatars.avatar"
  data-garden-version="version"
  disabled={true}
>
  <img
    src="test"
  />
</figure>
`;

exports[`Avatar applies default styling correctly 1`] = `
<figure
  className="c-avatar "
  data-garden-id="avatars.avatar"
  data-garden-version="version"
>
  <img
    src="test"
  />
</figure>
`;

exports[`Avatar applies system styling correclty if provided 1`] = `
<figure
  className="c-avatar c-avatar--system "
  data-garden-id="avatars.avatar"
  data-garden-version="version"
>
  <img
    src="test"
  />
</figure>
`;
